### DISCUSSION TITLE
Documentation Concept

### DISCUSSION THREAD

**Discussion by florianrusch-zf** [eclipse-edc/Connector#580] [Ideas]
We talked in a smaller group about how we can improve the documentation and made some notes.
Please review them and add comments if you like! Please also add comments if something is missing in the structure below!

## Target Groups

We identified two main groups of people who would benefit from a good documentation:

- "Technical Users" (which are just using the EDC) from companies which want to participant on a Data Space
- Developer
  - Contributor/Developer of the EDC Core
  - Extension-Developer

We write technical users in quotes because the name is not the best. It was chosen to clarify that e.g. managers are out of scope. For them, the general data space information in Gaia-X/Catena-X and on the Internet should be a better help. 

## Usage cases

We first tried to identify the use cases of the target groups in order to later create the structure of the documentation based on this.

Generell use-cases are:

- Understand how the EDC works
  - Architecture
  - Migration guide/Statement regarding the DSC (IDC)
- Get an overview of
  - all features
  - all extensions
- Build/Setup the connector
- How to interact with the connector
- Overview of samples

### Technical User

Additional to the general use-cases the technical user could also be interested in the following:

- Implement certain scenarios with the EDC framework
  - Monitoring
  - High Availability Setup
  - Infrastructure (k8s, cloud)
  - e.g. use the connector together with some corporate components/systems (ERP, DBs, Firewall, IAM, etc.)

### Contributor/Developer of the EDC and the extensions

- Setup a local environment to develop on the EDC framework
- Test the overall setup and functionality of the EDC framework to be sure, that new features don't break anything else
- Get an overview about the Patterns which are used by/in the EDC framework
- Extensions vs SPI's
- GitHub Workflow
- Contribution Guide

## Presentation

We have also discussed how the documentation should be presented. We currently have two main options that can be implemented with little or no effort:

- Markdown files in the repository (like currently)
- GitHub Pages
  - Docs folder
  - Separate branch (`gh-pages`)
  - Separate repository

For now, we would stay with the current process to avoid additional overhead.

## Structure

This is a draft of the structure and topics we would like to present in the documentation:

1. What is the EDC and how does it work
    1. Architecture
    1. Domain model
    1. Migration guide/Statement regarding the DSC (IDC) _[1]_
1. Overview of
    1. Features
    1. Extensions
1. **How to create a basic connector**
1. How to interact with the connector  _[2]_
1. **Overview of samples**
1. Operating a Connector
    1. Setup a connector with the EDC framework
    1. Topology/Architecture: How to interact with the connector _[2]_
    1. Implement certain scenarios with the EDC framework
        1. Monitoring
        1. High Availability
        1. Infrastructure (k8s, cloud)
        1. e.g. use the connector together with some corporate components/systems (ERP, DBs, Firewall, IAM, etc.)
1. Contribution & Development of the EDC
    1. Setup a local environment to develop on EDC framework
    1. **Test-Strategy**
        1. How to test single modules
        1. How to test the whole system
    1. Patterns _[3]_
        1. DataFlowController
        1. ...
    1. Extensions vs SPI's
    1. **GitHub Workflow**
    1. **Contribution Guide**
        1. Namings
            1. https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/discussions/553

In our eyes, the **bold** topics are either already documented and would only need to be revised, are quickly created, or could important for the first draft.

Some Notes:

1. Can probably be omitted since the EDC is to become standard. A first content could be taken from [`docs/ids/README.md`](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/26ab5bd3946b644035e74aae42a20cdcec515cd6/docs/ids/README.md)
2. I am not quite sure how this is different from each other.  I think, @MoritzKeppler, you have raised the second point. Perhaps you can bring some light into the darkness?
3. I am not very familiar with all the different patterns and techniques we are using in this project. Maybe some of the developers can help to fill this list?

## Next steps

- Let's discuss if the structure is okay or if something important is missing.
- Clarify prioritization (what to start with)
- Discuss how we can keep the documentation up2date
- Discuss how to get started

---
Edits:
- 31.01.2022 13:40
    - Change from bullet points to numbers in the structure so that changes are easier to identify
    - Add link to discussion regarding interface implementations (7. vi. a. a.)
- 22.02.2022 10:40
    - Add link to `docs/ids/README.md`

**Comment by dastaiger** [comment]
### Presentation of the Documentation
#### Template for html

[docsy](https://github.com/google/docsy/) is probably our best option.
It needs to be [set up](https://ruddra.com/hugo-deploy-static-page-using-github-actions/) with the use of GitHub actions. Otherwise it would require a manual step to generate the html documentation out of the .md files.


#### Templates für Issues und PRs inside of GitHub
[GitHub Templates](https://github.com/devspace/awesome-github-templates) to structure the interaction with the OpenSource community.
Might not be important right now, maybe something for the backlog

**Reply by florianrusch** [reply]
Thanks for the comment! Those are good points!

One additional comment about the presentation: if we want to publish the docs outside of the repo, we should also take a look at [mkdocs](https://github.com/mkdocs/mkdocs/). This project seems to be much more used than docsy (I also use it for a personal project).

Then later in the future we could go a step further and think about [vuepress](https://vuepress.vuejs.org/), which is the technology behind the [Vue.js documentation](https://vuejs.org/guide/quick-start.html). But for now, let's definitely start with something simple to drive progress.

**Reply by juliapampus** [reply]
With the DSC, we used Just the Docs (very basic, no huge style modifications): https://international-data-spaces-association.github.io/DataspaceConnector/. The "problem" with this is the metainformation that has to be added to each `.md` file to define paths and sortings etc. This is super easy but still extra effort. I would propose to use [docsify](https://github.com/docsifyjs/docsify). By adding at least 6 files (could be of course more if we decide on creating our own layout and style) and no additional metainformation to any file, the `/docs` folder is rendered directly. A sample setup can be seen here: https://fraunhoferisst.github.io/EclipseDataspaceConnector. What do you think? @florianrusch @florianrusch-zf

**Reply by juliapampus** [reply]
I am a fan of good UX and modern websites, so maybe this is not the prettiest solution, but by far the fastest and easiest. Without any unnecessary details. A search and toc is available, in my opinion you don't need more, considering that this is not supposed to be an official EDC website, but should make the docs slightly easier to access.

**Comment by florianrusch-zf** [comment]
## How we want to start improving the docs

First target group: "Technical Users"

General:
- Use some of the youtube videos @paullatzelsperger already made within the documentation (review them first)
- Focus at the architectural stuff
- Framework - "just checkout and start extending it"

Structure specific:
1. Overview about features and extensions. (for technical users who wants to identify if they can already use the edc with their feature sets they need)
   - Vault
   - AssetIndex
   - ... (the most common Extensions)
1. Write your own extension
1. How to setup certain scenarios/Use the EDC (e.g. AAS)

Regarding Features and extensions:
- Can we generate it from all README.md's?
- Template for a new extension: What should be documented?

Who works on which:
- @mspiekermann & @dastaiger Architectural Stuff
- @florianrusch-zf & @akelecevic Features & Extensions - https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/discussions/888
  - Speak with @alexandrudanciu
- @diegogomez-zf Domain Model - https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/pull/761
  - PPT from @MoritzKeppler
  - Excalidraw from @DominikPinsel

**Comment by juliapampus** [comment]
Hi all, while I was cleaning up the repo regarding milestones and projects, I stumbled across a few (closed) issues where Jim explains processes and implementations, some of them quite detailed. Maybe some of the content will help to avoid having to start from scratch.
- [ ] Issue #227 Contracts and Policy Integration 
- [ ] Issue #238 Data Loader
- [ ] Issue #239 Policy Negotiation
- [ ] Issue #463 Data Plane Framework
- [ ] Issue #541 Portable Transactions 
- [ ] Issue #622 Data Plane Framework
